File: internal/analyzers/plumbing/ticks_anomaly.go
package plumbing

import (
	"log"
	"sync/atomic"
	"time"
)

// minSaneCommitTime is the lower bound for a plausible committer timestamp.
// Git itself first shipped in 2005; commits stamped before 1990-01-01 are
// almost certainly the result of a corrupt commit object, an unset system
// clock (epoch 0 → 1970), or a deliberate `GIT_COMMITTER_DATE=` override.
//
// Without this clamp a single such commit pegged tick0 to ~1970, after
// which every modern commit's Sub(tick0) overflowed the int64-nanosecond
// [time.Duration] and clamped to ~292 years. That clamp leaked into burndown
// as a 106 740-day "analysis period". See ticks.go: the bug was sticky via
// max(tick, previousTick).
var minSaneCommitTime = time.Date(1990, time.January, 1, 0, 0, 0, 0, time.UTC)

// maxClockSkew is the upper-bound grace allowed past wall-clock time. A
// committer timestamp more than this far in the future is treated as
// anomalous regardless of repo content.
const maxClockSkew = 24 * time.Hour

// anomalyLogIntervalNanos throttles the per-event "anomalous committer
// timestamp" log line so a repo with thousands of bad commits doesn't
// drown the operator-facing log. Same shape as
// burndown/mismatch_tracker's log throttle.
const anomalyLogIntervalNanos = int64(time.Second)

// timeAnomalyTracker counts committer-timestamp anomalies detected during
// tick computation and rate-limits the warning log. Atomics make the
// tracker safe to call from the per-shard clones returned by Fork(); the
// sequential plumbing analyzer never actually races, but using atomics
// keeps Fork() safe by construction.
type timeAnomalyTracker struct {
	beforeMin    atomic.Int64 // Counter: timestamps before minSaneCommitTime.
	afterMax     atomic.Int64 // Counter: timestamps too far in the future.
	dropped      atomic.Int64 // Suppressed since last emitted log line.
	lastLogNanos atomic.Int64 // Monotonic-ish slot timestamp.
}

// recordBeforeMin bumps the before-min counter and emits a rate-limited
// warning. when is the bogus committer time we observed, replacement is
// the time we substituted into tick math.
func (t *timeAnomalyTracker) recordBeforeMin(when, replacement time.Time) {
	t.beforeMin.Add(1)
	t.maybeLog("before-min", when, replacement)
}

// recordAfterMax bumps the after-max counter and emits a rate-limited
// warning. Mirrors recordBeforeMin for the future-clamp side.
func (t *timeAnomalyTracker) recordAfterMax(when, replacement time.Time) {
	t.afterMax.Add(1)
	t.maybeLog("after-max", when, replacement)
}

// maybeLog emits one warning per anomalyLogIntervalNanos at most. Mirrors
// burndown.mismatchTracker.maybeLog: try to claim the slot via CAS; on
// failure (slot still warm), bump dropped and return silently. On success,
// flush the dropped tail in the emitted line.
func (t *timeAnomalyTracker) maybeLog(kind string, when, replacement time.Time) {
	now := time.Now().UnixNano()
	last := t.lastLogNanos.Load()

	if now-last < anomalyLogIntervalNanos {
		t.dropped.Add(1)

		return
	}

	if !t.lastLogNanos.CompareAndSwap(last, now) {
		t.dropped.Add(1)

		return
	}

	dropped := t.dropped.Swap(0)
	if dropped == 0 {
		log.Printf("ticks: %s anomalous committer timestamp %s, substituted %s",
			kind, when.Format(time.RFC3339), replacement.Format(time.RFC3339))

		return
	}

	log.Printf("ticks: %s anomalous committer timestamp %s, substituted %s [dropped=%d since last]",
		kind, when.Format(time.RFC3339), replacement.Format(time.RFC3339), dropped)
}

// snapshot returns the running counts. Used by accessor TimeAnomalies()
// for tests and external observers.
func (t *timeAnomalyTracker) snapshot() TimeAnomalyStats {
	return TimeAnomalyStats{
		BeforeMin: t.beforeMin.Load(),
		AfterMax:  t.afterMax.Load(),
	}
}

// TimeAnomalyStats reports anomalous committer-timestamp detections.
//
// BeforeMin counts commits whose committer time was earlier than the
// hard-coded floor (1990-01-01 UTC) — typically epoch-0 (1970) values
// from corrupt commit objects, unset system clocks, or deliberate
// GIT_COMMITTER_DATE overrides.
//
// AfterMax counts commits whose committer time was more than 24h past
// the analyzer's wall-clock — typically forged future timestamps
// ("--date=2099-01-01") or clock skew at commit time.
//
// In both cases the substituted time is the previous valid committer
// timestamp (or 1990-01-01 UTC if no valid commit has been seen yet),
// so the bad commit collapses onto the timeline at a sensible point
// instead of overflowing the int64-nanosecond Duration in ticks.go.
type TimeAnomalyStats struct {
	BeforeMin int64
	AfterMax  int64
}

// Total returns the combined count of anomalies on both bounds.
func (s TimeAnomalyStats) Total() int64 {
	return s.BeforeMin + s.AfterMax
}

File: rust/crates/cf-analyzers-plumbing/src/ticks_anomaly.rs
//! Committer-timestamp anomaly tracking.
//!
//! Port of `internal/analyzers/plumbing/ticks_anomaly.go`.
//!
//! Counts committer timestamps that fall outside the sane analysis window and
//! rate-limits the warning log, exactly mirroring the Go `timeAnomalyTracker`.

use std::sync::atomic::{AtomicI64, Ordering};
use std::sync::Arc;

/// Lower bound for a plausible committer timestamp, mirroring Go's
/// `minSaneCommitTime = time.Date(1990, 1, 1, 0, 0, 0, 0, time.UTC)`.
///
/// Expressed as a unix timestamp in seconds: 1990-01-01T00:00:00Z.
pub const MIN_SANE_COMMIT_TIME_UNIX: i64 = 631_152_000;

/// Upper-bound grace allowed past wall-clock time, mirroring Go's
/// `maxClockSkew = 24 * time.Hour`, in seconds.
pub const MAX_CLOCK_SKEW_SECS: i64 = 24 * 60 * 60;

/// Counts committer-timestamp anomalies, mirroring Go's `timeAnomalyTracker`.
///
/// The Go tracker rate-limits a warning log via atomics so the per-shard
/// `Fork()` clones stay safe by construction. The counters are the only part
/// observable through [`TimeAnomalyStats`]; the log line itself is operational
/// (not report) output, so its formatting parity is not byte-identity-binding,
/// but the counts are preserved exactly. Wrapped in an [`Arc`] so `Fork()`
/// clones share the aggregate, matching the Go field being a pointer shared
/// across clones.
#[derive(Debug, Default)]
pub struct TimeAnomalyTracker {
    before_min: AtomicI64,
    after_max: AtomicI64,
}

impl TimeAnomalyTracker {
    /// Construct an empty tracker.
    pub fn new() -> Arc<Self> {
        Arc::new(TimeAnomalyTracker::default())
    }

    /// Record a timestamp earlier than [`MIN_SANE_COMMIT_TIME_UNIX`].
    pub fn record_before_min(&self) {
        self.before_min.fetch_add(1, Ordering::Relaxed);
    }

    /// Record a timestamp too far in the future.
    pub fn record_after_max(&self) {
        self.after_max.fetch_add(1, Ordering::Relaxed);
    }

    /// Snapshot the running counts, mirroring Go's `snapshot()`.
    pub fn snapshot(&self) -> TimeAnomalyStats {
        TimeAnomalyStats {
            before_min: self.before_min.load(Ordering::Relaxed),
            after_max: self.after_max.load(Ordering::Relaxed),
        }
    }
}

/// Anomalous committer-timestamp detections, mirroring Go's `TimeAnomalyStats`.
#[derive(Debug, Clone, Copy, Default, PartialEq, Eq)]
pub struct TimeAnomalyStats {
    /// Count of timestamps earlier than 1990-01-01 UTC.
    pub before_min: i64,
    /// Count of timestamps more than `maxClockSkew` past wall-clock.
    pub after_max: i64,
}

impl TimeAnomalyStats {
    /// Combined count of anomalies on both bounds, mirroring Go's `Total()`.
    pub fn total(&self) -> i64 {
        self.before_min + self.after_max
    }
}
